File: python/flyql/generators/postgresql/column.py
"""PostgreSQL-dialect Column."""

import re
from typing import Dict, List, Optional

from flyql.flyql_type import Type

REGEX = {
    Type.String: re.compile(
        r"^(varchar|char|character varying|character)\s*\(\s*\d+\s*\)", re.IGNORECASE
    ),
    Type.Float: re.compile(
        r"^(numeric|decimal)\s*\(\s*\d+\s*(,\s*\d+)?\s*\)", re.IGNORECASE
    ),
    Type.Date: re.compile(r"^timestamp\s*\(\s*\d+\s*\)", re.IGNORECASE),
    Type.Array: re.compile(r"(\[\]$|^_)", re.IGNORECASE),
}

# PG renames: jsonb→json, hstore→map, interval→duration.
FLYQL_TYPE_TO_POSTGRESQL_TYPES: Dict[Type, set] = {
    Type.String: {
        "text",
        "varchar",
        "char",
        "character varying",
        "character",
        "name",
        "uuid",
        "citext",
        "inet",
        "cidr",
        "macaddr",
    },
    Type.Int: {
        "smallint",
        "integer",
        "bigint",
        "int2",
        "int4",
        "int8",
        "serial",
        "bigserial",
        "smallserial",
    },
    Type.Float: {
        "real",
        "double precision",
        "numeric",
        "decimal",
        "float4",
        "float8",
        "money",
    },
    Type.Bool: {"boolean", "bool"},
    Type.Date: {
        "date",
        "timestamp",
        "timestamptz",
        "timestamp without time zone",
        "timestamp with time zone",
        "time",
        "timetz",
    },
    Type.Duration: {"interval"},
    Type.JSON: {"jsonb", "json"},
    Type.Map: {"hstore"},
}


def normalize_postgresql_type(pg_type: str) -> Type:
    if not pg_type or not isinstance(pg_type, str):
        return Type.Unknown

    normalized = pg_type.strip().lower()

    if normalized == "jsonstring":
        return Type.JSONString

    if REGEX[Type.Array].search(normalized):
        return Type.Array

    if REGEX[Type.String].match(normalized):
        return Type.String
    if normalized in FLYQL_TYPE_TO_POSTGRESQL_TYPES[Type.String]:
        return Type.String

    if normalized in FLYQL_TYPE_TO_POSTGRESQL_TYPES[Type.Int]:
        return Type.Int

    if REGEX[Type.Float].match(normalized):
        return Type.Float
    if normalized in FLYQL_TYPE_TO_POSTGRESQL_TYPES[Type.Float]:
        return Type.Float

    if normalized in FLYQL_TYPE_TO_POSTGRESQL_TYPES[Type.Bool]:
        return Type.Bool

    if REGEX[Type.Date].match(normalized):
        return Type.Date
    if normalized in FLYQL_TYPE_TO_POSTGRESQL_TYPES[Type.Date]:
        return Type.Date

    if normalized in FLYQL_TYPE_TO_POSTGRESQL_TYPES[Type.Duration]:
        return Type.Duration

    if normalized in FLYQL_TYPE_TO_POSTGRESQL_TYPES[Type.JSON]:
        return Type.JSON

    if normalized in FLYQL_TYPE_TO_POSTGRESQL_TYPES[Type.Map]:
        return Type.Map

    return Type.Unknown


class Column:
    """Opaque PostgreSQL-dialect column."""

    def __init__(
        self,
        name: str,
        _type: str,
        values: Optional[List[str]] = None,
        display_name: str = "",
        raw_identifier: str = "",
    ):
        self.name = name
        self.match_name = name
        self.values: List[str] = values or []
        self.display_name = display_name
        self.raw_identifier = raw_identifier
        self._raw_type = _type
        self._flyql_type: Type = normalize_postgresql_type(_type)

    @property
    def raw_type(self) -> str:
        return self._raw_type

    @property
    def flyql_type(self) -> Type:
        return self._flyql_type

    def with_raw_identifier(self, identifier: str) -> "Column":
        self.raw_identifier = identifier
        return self

File: javascript/packages/flyql/src/generators/postgresql/column.js
import { Type } from '../../flyql_type.js'
import { FlyqlError } from '../../core/exceptions.js'

const typeRegexes = {
    [Type.String]: /^(varchar|char|character varying|character)\s*\(\s*\d+\s*\)/i,
    [Type.Float]: /^(numeric|decimal)\s*\(\s*\d+\s*(,\s*\d+)?\s*\)/i,
    [Type.Date]: /^timestamp\s*\(\s*\d+\s*\)/i,
    [Type.Array]: /(\[\]$|^_)/i,
}

const flyqlTypeToPostgreSQLTypes = {
    [Type.String]: new Set([
        'text',
        'varchar',
        'char',
        'character varying',
        'character',
        'name',
        'uuid',
        'citext',
        'inet',
        'cidr',
        'macaddr',
    ]),
    [Type.Int]: new Set([
        'smallint',
        'integer',
        'bigint',
        'int2',
        'int4',
        'int8',
        'serial',
        'bigserial',
        'smallserial',
    ]),
    [Type.Float]: new Set(['real', 'double precision', 'numeric', 'decimal', 'float4', 'float8', 'money']),
    [Type.Bool]: new Set(['boolean', 'bool']),
    [Type.Date]: new Set([
        'date',
        'timestamp',
        'timestamptz',
        'timestamp without time zone',
        'timestamp with time zone',
        'time',
        'timetz',
    ]),
    [Type.Duration]: new Set(['interval']),
    [Type.JSON]: new Set(['jsonb', 'json']),
    [Type.Map]: new Set(['hstore']),
}

export function normalizePostgreSQLType(pgType) {
    if (!pgType) return Type.Unknown

    const normalized = pgType.trim().toLowerCase()

    if (normalized === 'jsonstring') return Type.JSONString

    if (typeRegexes[Type.Array].test(normalized)) return Type.Array

    if (typeRegexes[Type.String].test(normalized)) return Type.String
    if (flyqlTypeToPostgreSQLTypes[Type.String].has(normalized)) return Type.String

    if (flyqlTypeToPostgreSQLTypes[Type.Int].has(normalized)) return Type.Int

    if (typeRegexes[Type.Float].test(normalized)) return Type.Float
    if (flyqlTypeToPostgreSQLTypes[Type.Float].has(normalized)) return Type.Float

    if (flyqlTypeToPostgreSQLTypes[Type.Bool].has(normalized)) return Type.Bool

    if (typeRegexes[Type.Date].test(normalized)) return Type.Date
    if (flyqlTypeToPostgreSQLTypes[Type.Date].has(normalized)) return Type.Date

    if (flyqlTypeToPostgreSQLTypes[Type.Duration].has(normalized)) return Type.Duration

    if (flyqlTypeToPostgreSQLTypes[Type.JSON].has(normalized)) return Type.JSON

    if (flyqlTypeToPostgreSQLTypes[Type.Map].has(normalized)) return Type.Map

    return Type.Unknown
}

/**
 * Opaque PostgreSQL-dialect Column. Construct via `newColumn({...})` or `new Column({...})`.
 * The flyql semantic type is computed at construction via `normalizePostgreSQLType`.
 */
export class Column {
    constructor(opts) {
        if (opts == null || typeof opts !== 'object' || Array.isArray(opts)) {
            throw new FlyqlError(`Column: expected an options object ({ name, type, values? }); got ${typeof opts}`)
        }
        const { name, type, values = null, displayName = '', rawIdentifier = '' } = opts
        if (typeof name !== 'string' || !name) {
            throw new FlyqlError(`Column: 'name' must be a non-empty string`)
        }
        if (typeof type !== 'string') {
            throw new FlyqlError(
                `Column: 'type' must be a raw-type string (e.g. 'text', 'jsonstring'); got ${typeof type}`,
            )
        }

        this.name = name
        this.values = values || []
        this.displayName = displayName
        this.rawIdentifier = rawIdentifier
        this.matchName = name
        this._rawType = type
        this._flyqlType = normalizePostgreSQLType(type)
    }

    rawType() {
        return this._rawType
    }

    flyqlType() {
        return this._flyqlType
    }

    /** @deprecated Pass `rawIdentifier` in the options object to `newColumn()` / `new Column()` instead. Slated for removal in a follow-up. */
    withRawIdentifier(identifier) {
        this.rawIdentifier = identifier
        return this
    }
}

export function newColumn(opts) {
    if (opts == null || typeof opts !== 'object' || Array.isArray(opts)) {
        throw new FlyqlError(`newColumn: expected an options object ({ name, type, values? }); got ${typeof opts}`)
    }
    return new Column(opts)
}
